### ISSUE TITLE
Feature request: Escaping { and } in url

### ISSUE THREAD

**Issue by oliverhausler** [square/retrofit#754]
I am using retrofit to access the Facebook API. Because Facebook uses curly braces {} within their query syntax (and retrofit does too), it is quite confusing to write against the Facebook API.

https://developers.facebook.com/docs/graph-api/using-graph-api/v2.2#fieldexpansion

This is how a Facebook request looks like:

    @GET("/me?fields=id,email,first_name,gender,last_name,link,locale,name,timezone,verified,picture.type(large){is_silhouette,url},age_range") JsonObject getUser(@Query("access_token") String userToken);

But because retrofit expects a parameter within {} I need to escape the curly braces meant for Facebook (which works, but looks kind of ugly and is hard to read):

    @GET("/me?fields=id,email,first_name,gender,last_name,link,locale,name,timezone,verified,picture.type(large)%7Bis_silhouette,url%7D,age_range") JsonObject getUser(@Query("access_token") String userToken);

Suggestion:

If you don't see any negative side effects, I would suggest allowing to escape { with {{ and } with }} if the curly braces are meant for the API and not for retrofit.

You could run a simple pre-check which replaces {{ with %7B and }} with %7D and the user would have a better readable endpoint definition:

    @GET("/me?fields=id,email,first_name,gender,last_name,link,locale,name,timezone,verified,picture.type(large){{is_silhouette,url}},age_range") JsonObject getUser(@Query("access_token") String userToken);

I admit it's a special use-case, but Facebook is not nobody ;-)

**Comment by vignesh-iopex** [comment]
I find that the Retrofit processes from annotation parameters to the matching relative url {string}. So If don't have the `@Path` annotation for the url substring `{is_silhouette,url}` Retrofit wouldn't process the string for replacement.

**Comment by oliverhausler** [comment]
Might be a language problem, but I don't understand, sorry. Would you please say that again in other words, a bit more detailed.

**Comment by vignesh-iopex** [comment]
Retrofit looks for every `@Path("substring")` annotations and replaces respective `{substring}` from the relative url so If we don't have the `@Path` annotation for the braces enclosed substring it won't replace them.

Have you tried using the braces in the url but not having `@Path` annotations ?

**Comment by oliverhausler** [comment]
Facebook uses the braces as part of their own URL path and the problem is that Retrofit replaces them or raises an exception because it thinks the braces are meant for itself. The Facebook URL officially contains braces, which clashes with the Retrofit annotation parameter. So my suggestion was for Retrofit to leave content in double braces {{content}} alone and convert the double braces to single braces {content}.

In other words: single braces are for Retrofit, double braces are pass-through.

**Comment by oliverhausler** [comment]
We could also use Java-style escape, such as \{ and \} to tell Retrofit to ignore a brace. Probably cleaner than {{ or }} if \ does not cause problems otherwise.
